nautilus_trader/backtest/config.py

```diff
@@ -0,0 +1,277 @@
+        self._check_kwargs(kwargs)
+        self.__dict__.update(kwargs)
+        return self
+
+    def replace(self, **kwargs):
+        """Return a new instance with some attributes replaces"""
+        return self.__class__(
+            **{**{k: getattr(self, k) for k in self.__dataclass_fields__}, **kwargs}
+        )
+
+    def __repr__(self):
+        dataclass_repr_func = dataclasses._repr_fn(
+            fields=list(self.__dataclass_fields__.values()), globals=self.__dict__
+        )
+        r = dataclass_repr_func(self)
+        if self.missing():
+            return "Partial-" + r
+        return r
+
+
+@dataclasses.dataclass()
+class BacktestDataConfig:
+    data_type: type
+    instrument_id: Optional[str] = None
+    start_time: Optional[int] = None
+    end_time: Optional[int] = None
+    filters: Optional[dict] = None
+    client_id: Optional[str] = None
+
+    @property
+    def query(self):
+        return dict(
+            cls=self.data_type,
+            instrument_ids=[self.instrument_id] if self.instrument_id else None,
+            start=self.start_time,
+            end=self.end_time,
+            as_nautilus=True,
+        )
+
+
+@dataclasses.dataclass()
+class BacktestVenueConfig:
+    name: str
+    venue_type: str
+    oms_type: str
+    account_type: str
+    base_currency: Currency
+    starting_balances: List[Money]
+    fill_model: Optional[FillModel] = None
+    modules: Optional[List[SimulationModule]] = None
+
+    def __dask_tokenize__(self):
+        values = [
+            self.name,
+            self.venue_type,
+            self.oms_type,
+            self.account_type,
+            self.base_currency.code,
+            ",".join(sorted([balance.to_str() for balance in self.starting_balances])),
+            self.modules,
+        ]
+        return tuple(values)
+
+
+@dataclasses.dataclass(repr=False)
+class BacktestConfig(Partialable):
+    """
+    Configuration for one specific backtest run (a single set of data / strategies / parameters)
+    """
+
+    venues: Optional[List[BacktestVenueConfig]] = None
+    instruments: Optional[List[Instrument]] = None
+    data_config: Optional[List[BacktestDataConfig]] = None
+    strategies: Optional[List[Tuple[type, dict]]] = None
+    name: Optional[str] = None
+    # data_catalog_path: Optional[str] = None
+
+
+def _load(config: BacktestDataConfig):
+    catalog = DataCatalog()
+    query = config.query
+    return {
+        "type": query["cls"],
+        "data": catalog.query(**query),
+        "client_id": config.client_id,
+    }
+
+
+@delayed(pure=True)
+def load(config: BacktestDataConfig):
+    return _load(config=config)
+
+
+# @delayed(pure=True)
+def create_backtest_engine(venues, instruments, data):
+    engine = BacktestEngine(
+        bypass_logging=True,
+        run_analysis=False,
+    )
+
+    # Add Instruments
+    for instrument in instruments:
+        engine.add_instrument(instrument)
+
+    # Add data
+    for d in data:
```

Comment:
Is there a way for us to pass `DataFrames` to the engine?

This is how I've been doing it, maybe theres a better way:

```python
engine.add_instrument(instrument_provider)
engine.add_bars(instrument_provider.id, BarAggregation.MINUTE, PriceType.BID, df_bid_data)
engine.add_bars(instrument_provider.id, BarAggregation.MINUTE, PriceType.ASK, df_ask_data)
```

Reply:
Yeah I've been thinking about this a little more the past few days while playing with this PR. Originally I thought it would simplify the code if we could funnel all sources into nautilus objects (simplifying the backtest engine), however I'm now leaning towards using dataframes more centrally for the following reasons:
- We store data in the catalog in parquet format (dataframes basically).
- Dask has much better support for passing dataframes around a cluster (serialization) than custom python or even cython objects.  

The only question/issue with this approach is how quickly we can create the objects to load up the engine. The benefit to using nautilus objects is that the loading/parsing can be done in another worker and sent to the engine process (mostly IO in a background thread). We probably need to do some testing to see how much this would slow down the engine runs.

I'm hoping to do some of this work in the coming weeks, but would welcome any thoughts or contributions!

Reply:
What do you mean by `nautilus objects`? Wondering if converting DataFrames to a nautilus object would be useful. In my limited knowledge and testing, for the same data range, loading it through the CSV loader in this PR is 2x faster than when passing the DataFrame directly.

>The only question/issue with this approach is how quickly we can create the objects to load up the engine. 

At first I tried to modify the DataLoader so it would be seamless, but kept running into errors, so opted to make it work initially. These are my local changes to get it working (again this method is 2x slower =/). By 2x slower, I tested it by creating 98 unique parameter sets and running 14 Dask workers using CSV loading method + DataFrame loading method:
```python
PARAM_SET = []
for p in range(10, 500, 5):
    PARAM_SET.append({"fast_ema": p, "slow_ema": p * 2})
```

DataFrame loading method I used:

1. Copied your method `parse_csv_tick` and refactored it to parse a `DataFrame`:

```python
def nt_df_to_quote_tick(df, instrument):
    qt = []
    for r in df[['date', 'bid_close', 'ask_close']].values:
        ts = secs_to_nanos(pd.Timestamp(r[0]).timestamp())
        tick = QuoteTick(
            instrument_id=instrument.id,
            bid=Price.from_str(str(r[1])),
            ask=Price.from_str(str(r[2])),
            bid_size=Quantity.from_int(1_000_000),
            ask_size=Quantity.from_int(1_000_000),
            ts_event_ns=ts,
            ts_recv_ns=ts,
        )
        qt.append(tick)

    return qt
```

2. Load it into the config
```python
    config = base.update(
        instruments=[instrument],
        data_config={"data": qt_ticks},
    )
```

3. To load it directly, I modified https://github.com/nautechsystems/nautilus_trader/pull/361/files#diff-5032d40ebe929430aa62c7a115998d23b10309b7549833cb56820717d30d73e6#R250-R260 

```python
def build_graph(backtest_configs, sync=False, custom_data=False):
    backtest_configs = _check_configs(backtest_configs)

    _ = (
        DataCatalog()
    )

    results = []
    for config in backtest_configs:
        config.check(ignore=("name",))
        input_data = []
        for data_config in config.data_config:
            # Added if statement here to load it
            if custom_data:
                input_data.append({"data": config.data_config[data_config], "type": QuoteTick, "client_id": None})
            else:
                load_func = (
                    _load
                    if sync
                    else partial(load, dask_key_name=f"load-{tokenize(data_config.query)}")
                )
```

Reply:
By `nautilus_objects` I mean any of the models within nautilus (QuoteTick, TradeTick) in their actual python `object` instance form (not a dataframe or csv). 

I think what you've got above is fine, but it's going to be (much) slower if you need to load more data. Say for example you are backtesting a single instrument (say BTC) over a longer period of time (1 year). I believe you will likely spend a large % of your backtest total time in loading and parsing the data (and a small % in actually running the engine). 

I'm going to create a notebook with some rough benchmarks we can discuss. Will come back soon